### DISCUSSION TITLE
How to prevent `nuxt dev` from calling useAsyncData?

### DISCUSSION THREAD

**Discussion by ghost** [nuxt/nuxt#18894] [Questions]
I am fetching data using `useAsyncData` from an API with a secret token which is stored in my `.env` file. In `nuxt.config.js` I have defined that secret token as private `runtimeConfig`.

This works great for SSG with `nuxt generate`, but in dev mode with `nuxt dev` it does not work as expected.

So in dev mode, when I load the page everthing works fine, because the data is fetched server side (inside useAsyncData). But when navigating to a different route client side, I will get an error because `useAsyncData` is called and nuxt tries to load the private config var — which does not exist on the client side.

So because I only need `nuxt dev` in dev mode, I was wondering how I can make it work in dev mode too?

Any ideas?

**Comment by AloisSeckar** [comment]
I think, you can put your data fetching inside `/server` directory, because this is server side, where private config is available. Then from client-side context you don't make your fetch call directly, but you call your own server route and do the real data fetch here with your secret.

I created this to demonstrate the idea - you can access private config in `/server/api/secret.get.ts` and you can even expose it to client's side (OFC it is not a good idea to do so)
https://stackblitz.com/edit/nuxt-starter-696yxp?file=pages/index.vue

**Reply by ghost** [reply]
@AloisSeckar Thanks for your great demo! Oh well, this works with `nuxt dev` but not with `nuxt generate` or `npx nuxi generate && npx serve .output/public`, because we can’t access `/server/api` in SSG mode, I guess. [Or can we?](https://github.com/nuxt/nuxt/discussions/16068) I mean it would be amazing.

**Reply by AloisSeckar** [reply]
My bad. I didnt realize, you need something that will work in both `nuxt dev` and SSG. I am afraid I cannot see, how server routes, dynamic by nature, could be converted into static output :(

But back to your use case - would be a problem to have another testing token, that could be exposed as public in your dev environment? Then you can check whether the public token is set - if yes (in DEV), use it, if not, go for the private one (only in PROD).

With a composable like that, you should easilly determine prior to any API call:

```js
function getToken() {
  const config = useRuntimeConfig()
  if (config.public.token) {
    return config.public.token
  } else {
    return config.token
  }
}
```

Or do I miss something again?

**Reply by ghost** [reply]
@AloisSeckar Thanks. That’s exactly what I am doing at the moment, but I was wondering if there is a built in solution for this case …

> I am afraid I cannot see, how server routes, dynamic by nature, could be converted into static output :(

It’s more about providing server routes than converting them — but feel free to [join the discussion](https://github.com/nuxt/nuxt/discussions/16068).
